File: frontends/tackle_cli/lib/tackle_cli/widgets/browse.ex
defmodule Tackle.CLI.Widgets.Browse do
  @moduledoc """
  Tackle-owned native Browse viewport. Rust paints responsive page tabs, wraps
  plain diagnostic documents, clamps scrolling and clips visible rows. Transcript
  pages reuse immutable Conversation resources and their existing selection.
  Elixir retains source for copying and owns page selection and runtime data.
  Resources are local, immutable, and must not be persisted or shared across nodes.
  """

  alias ExRatatui.Layout.Rect
  alias Tackle.CLI.Native
  alias Tackle.CLI.TUI.Theme
  alias Tackle.CLI.Widgets.{Conversation, Surface}

  defstruct [:state, page: 0, scroll_offset: 0, selected: []]

  @type t :: %__MODULE__{
          state: reference(),
          page: non_neg_integer(),
          scroll_offset: non_neg_integer(),
          selected: [non_neg_integer()]
        }

  @doc "Measures a frozen plain-text document at the given width."
  @spec document(String.t(), pos_integer()) :: {reference(), non_neg_integer()}
  def document(text, width),
    do: Native.browse_document(text, width, Conversation.style(Theme.style(:text)))

  @doc "Clamps row movement using the native viewport's chrome and content height."
  @spec scroll(non_neg_integer(), non_neg_integer(), non_neg_integer(), integer()) ::
          non_neg_integer()
  defdelegate scroll(offset, total, height, delta), to: Native, as: :browse_scroll

  @doc "Renders tabs and the selected page through the owned surface wire format."
  @spec render(t(), Rect.t()) :: [{struct(), Rect.t()}]
  def render(widget, rect) do
    styles =
      {Conversation.style(Theme.style(:accent)), Conversation.style(Theme.style(:muted)),
       Conversation.style(Theme.style(:selection_surface))}

    {:ok, rows} =
      Native.browse_render(
        widget.state,
        widget.page,
        rect.width,
        rect.height,
        widget.scroll_offset,
        widget.selected,
        styles
      )

    Surface.place(rows, rect)
  end

  defimpl ExRatatui.Widget do
    defdelegate render(widget, rect), to: Tackle.CLI.Widgets.Browse
  end
end

File: apps/tackle_cli/native/tackle/src/browse.rs
//! Immutable local Browse resources. No provider or session state enters Rust.
use crate::{
    atoms,
    conversation::{style, ConversationResource, WireStyle},
    surface,
    widgets::{
        browse::{body_height, scroll, Browse, PAGES},
        conversation::{Conversation, HistoryCell},
    },
    MAX_CELLS,
};
use ratatui::{buffer::Buffer, layout::Rect, widgets::Widget};
use rustler::{Atom, Error, NifResult, ResourceArc};
use std::sync::Arc;

#[rustler::nif(schedule = "DirtyCpu")]
fn browse_document(
    source: String,
    width: u16,
    base: WireStyle,
) -> NifResult<(ResourceArc<ConversationResource>, usize)> {
    let cell = HistoryCell::plain(&source, width, style(base)?);
    let conversation = Conversation::new(vec![Arc::new(cell)]);
    let height = conversation.height;
    Ok((
        ResourceArc::new(ConversationResource(conversation, width.max(1))),
        height,
    ))
}

#[rustler::nif]
fn browse_scroll(offset: usize, total: usize, height: u16, delta: i64) -> usize {
    scroll(offset, total, body_height(height), delta)
}

#[rustler::nif(schedule = "DirtyCpu")]
fn browse_render(
    resource: ResourceArc<ConversationResource>,
    page: usize,
    width: u16,
    height: u16,
    offset: usize,
    selected: Vec<usize>,
    styles: (WireStyle, WireStyle, WireStyle),
) -> NifResult<(Atom, Vec<surface::Line>)> {
    if usize::from(width) * usize::from(height) > MAX_CELLS || selected.len() > MAX_CELLS {
        return Err(Error::Term(Box::new(atoms::invalid_size())));
    }
    if page >= PAGES.len() || (width != 0 && height != 0 && width != resource.1) {
        return Err(Error::BadArg);
    }
    let area = Rect::new(0, 0, width, height);
    let mut buffer = Buffer::empty(area);
    Browse {
        conversation: &resource.0,
        page,
        offset,
        selected: &selected,
        accent: style(styles.0)?,
        muted: style(styles.1)?,
        selection: style(styles.2)?,
    }
    .render(area, &mut buffer);
    let rows = surface::lines(&buffer).map_err(|reason| Error::Term(Box::new(reason)))?;
    Ok((atoms::ok(), rows))
}
